### PR TITLE
MOD-7214: Avoid using redis to retrieve a value of sortable field if value should be in cache and is not(i.e missing)

### PR DESCRIPTION
**Pull request by kei-nan**
A clear and concise description of what the PR is solving, including:
1. The current state briefly - we kept accessing redis modules api to retrieve the value of a field even when we didn't need to.
2. What is the change
* allow setting forceLoad from the caller, hopefully this will allow the caching optimization to be used more often when loading keys from loadIndividualKeys
* If key is sortable and value should be present in the cache and we get null then we can assume value is the same as the document.
3. Adding the outcome (changed state)
We should now only use redis modules api when needed

**Which issues this PR fixes**
1. [MOD-7214](https://redislabs.atlassian.net/browse/MOD-7214)

**Main objects this PR modified**
1. Loader

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

[MOD-7214]: https://redislabs.atlassian.net/browse/MOD-7214?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by raz-mon — CHANGES_REQUESTED** [review]
Nice work!
A couple of comments:
1. Let's add tests for the missing values indexing in the aggregation pipeline.
2. Can we add a test validating that we don't use Redis when we don't intend to? For instance in the case of missing indexing of a sortable field.
At the end we call the `HGET` or `HGETALL` command, which is probably counted in the command statistics - that may be a good indicator. This may be a nice general test to add as well, to prevent situations like the one we may be in, in which we think we don't call Redis (which is the whole point of our cache), but we actually do.

File: src/rlookup.c [hunk 1]
```diff
@@ -599,10 +599,25 @@ static RSValue *replyElemToValue(RedisModuleCallReply *rep, RLookupCoerceType ot
   }
 }
 
+static bool isValueAvailable(const RLookupKey *kk, const RLookupRow *dst, RLookupLoadOptions *options)
+{
+  if (!options->forceLoad && kk && !(kk->flags & RLOOKUP_F_FORCE_LOAD)) {
```
Comment (GuyAv46) [inline src/rlookup.c:604]: I don't think we need to validate `kk` here

1. I don't think it's ever NULL
2. If it was NULL, this condition yields `false`, and the next condition (else if) will dereference it and we will crash

Comment (GuyAv46) [inline src/rlookup.c:604]: `RLOOKUP_F_FORCE_LOAD` is never turned on on a key (see `RLOOKUP_TRANSIENT_FLAGS` definition)

File: src/rlookup.c [hunk 2]
```diff
@@ -962,7 +976,6 @@ int RLookup_LoadRuleFields(RedisModuleCtx *ctx, RLookup *it, RLookupRow *dst, In
                             .keyPtr = keyptr,
                             .type = rule->type,
                             .status = &status,
-                            .forceLoad = 1,
```
[the remark this answers is not in the document]
Comment (kei-nan) [reply src/rlookup.c:965]: It is needed, I think it is remainder of when I tried to remove forceLoad altogether.

File: src/rlookup.c [hunk 3]
```diff
@@ -599,10 +599,24 @@ static RSValue *replyElemToValue(RedisModuleCallReply *rep, RLookupCoerceType ot
   }
 }
 
+static bool isValueAvailable(const RLookupKey *kk, const RLookupRow *dst, RLookupLoadOptions *options) {
+  if (!options->forceLoad) {
+    if (kk->flags & RLOOKUP_F_VAL_AVAILABLE) {
+      // No need to "write" this key. It's always implicitly loaded!
+      return true;
+    } else if ((kk->flags & (RLOOKUP_F_SVSRC | RLOOKUP_F_ISLOADED)) == (RLOOKUP_F_SVSRC | RLOOKUP_F_ISLOADED)) {
```
Comment (GuyAv46) [inline src/rlookup.c:607]: Why do we need to check if the key `RLOOKUP_F_ISLOADED`?

1. `RLOOKUP_F_SVSRC` and `RLookup_GetItem(kk, dst) == NULL` is enough evidence we don't need to load the value right?
    - have we checked what is the normalized form of an empty string, a string with only whitespace? If it's NULL we might have a problem
3. We shouldn't be here if the key is marked with `RLOOKUP_F_ISLOADED`. It means that a loader from the upstream already loaded the value, so following loaders that load individual keys shouldn't try to load it again.
    - see `(key->flags & RLOOKUP_F_ISLOADED &&       !(flags &  RLOOKUP_F_OVERRIDE))` condition in `GetKey`

**Comment by kei-nan** [comment]
Hi @raz-mon,
We use RedisModule_ScanKey,
I saw the command stats but do we have API usage stats?

```
127.0.0.1:6379> FLUSHALL
OK
127.0.0.1:6379> ft.create idx schema foo text sortable
OK
127.0.0.1:6379> hset doc foo hello
(integer) 1
127.0.0.1:6379> FT.SEARCH idx *
1) (integer) 1
2) "doc"
3) 1) "foo"
   2) "hello"
127.0.0.1:6379> INFO COMMANDSTATS
# Commandstats
cmdstat_flushall:calls=1,usec=4714,usec_per_call=4714.00,rejected_calls=0,failed_calls=0
cmdstat_FT.SEARCH:calls=1,usec=163,usec_per_call=163.00,rejected_calls=0,failed_calls=0
cmdstat_hset:calls=1,usec=773,usec_per_call=773.00,rejected_calls=0,failed_calls=0
cmdstat_FT.CREATE:calls=2,usec=563,usec_per_call=281.50,rejected_calls=0,failed_calls=1
cmdstat_info:calls=1,usec=15,usec_per_call=15.00,rejected_calls=0,failed_calls=0
cmdstat_command|docs:calls=1,usec=1699,usec_per_call=1699.00,rejected_calls=0,failed_calls=0

```

File: src/rlookup.c [hunk 4]
```diff
@@ -599,10 +599,22 @@ static RSValue *replyElemToValue(RedisModuleCallReply *rep, RLookupCoerceType ot
...
+  if (!options->forceLoad) {
+    if (kk->flags & RLOOKUP_F_VAL_AVAILABLE) {
+      // No need to "write" this key. It's always implicitly loaded!
+      return true;
+    } else if ((kk->flags & RLOOKUP_F_SVSRC) && (RLookup_GetItem(kk, dst) == NULL)) {
+      // There is no value in the sorting vector, and we don't need to load it from the document.
+      return true;
+    }
+  }
+  return false;
```
Comment (GuyAv46) [inline src/rlookup.c:612]: Consider avoiding `if (cond) return true; else return false;` pattern
```suggestion
  return (!options->forceLoad && (
      // No need to "write" this key. It's always implicitly loaded!
      (kk->flags & RLOOKUP_F_VAL_AVAILABLE) ||
      // There is no value in the sorting vector, and we don't need to load it from the document.
      ((kk->flags & RLOOKUP_F_SVSRC) && (RLookup_GetItem(kk, dst) == NULL))
  ));
```

File: src/rlookup.c [hunk 5]
```diff
@@ -599,10 +599,21 @@ static RSValue *replyElemToValue(RedisModuleCallReply *rep, RLookupCoerceType ot
   }
 }
 
+// returns trus if the value of the key is already available
```
Comment (GuyAv46) [inline src/rlookup.c:602]:
```suggestion
// returns true if the value of the key is already available
```

File: src/rlookup.c [hunk 6]
```diff
@@ -599,6 +599,18 @@ static RSValue *replyElemToValue(RedisModuleCallReply *rep, RLookupCoerceType ot
   }
 }
 
+// returns true if the value of the key is already available
+// avoids the need to call to redis api to get the value
+// i.e we can use the sorting vector as a cache
+static bool isValueAvailable(const RLookupKey *kk, const RLookupRow *dst, RLookupLoadOptions *options) {
```
Comment (GuyAv46) [inline src/rlookup.c:605]: Why don't we use this optimization?

File: src/rlookup.c [hunk 7]
```diff
@@ -599,10 +599,21 @@ static RSValue *replyElemToValue(RedisModuleCallReply *rep, RLookupCoerceType ot
...
```
Comment (GuyAv46) [inline src/rlookup.c:605]: consider making `inline`

**Review by raz-mon — COMMENTED** [review]
LGTM in general, nice work.
Two questions that I would love to understand: 
1. How many of the cases in which we should have avoided querying redis yet still queried it are we now avoiding? 
For that:
* please run the benchmarks for this PR.
* Do we have a way to know if we called the Redis API or not? If we do - it can be a great addition to our tests. Specifically, we can from now on validate that we have not reached Redis when we don't want to - something that should have been added a long time ago.

2. Since we saw some discrepancy in the results we return from Redis (via `HGETALL`) and from our sorting vector, can this modification change the type of a result for a specific scenario (data + query)? Specifically we saw a hazard in the `numeric` case, so that would be a good place to check. If the answer to this is yes - we should keep the current `forceLoad=1` and unfortunately give up on this optimization until we can break (i.e., merge the part changing the initialization of `forceLoad` to master only, and keep only the optimization for the missing case here - which is already a very nice optimization!).

**Comment by kei-nan** [comment]
https://github.com/RediSearch/RediSearch/pull/4729#pullrequestreview-2123346475
1. 
I can run the benchmarks tomorrow.
I am not aware of any way to know when we called redis API or not, maybe there is a tool somewhere that uses hooks to see how many extern c function calls were performed, doesn't this part feels a bit out of scope for this PR?

2.
There is an existing discrepancy in the types between the flow of HGETALL, loadIndividualKeys and RLookup_GetKey_LoadEx.
- For HGETALL - we changed it so it would implicity force load everything -> behaves as it did previously
- For loadIndividualKeys - Now we can return OK for unf fields as was originally intended by no longer setting forceLoad to 1.
  - The main thing to note is that in cases where a loader is not needed and only RLookup_GetKey_Load is used then the same caching behavior is used by returning null based on the key flags.
  - But notice that in that flow we don't force the type to be a string anyhow
  - That means that the type remains the same in either loadIndividualKeys or RLookup_GetKey_LoadEx
  - So type wise there shouldn't be a change.

**Comment by raz-mon** [comment]
@kei-nan Yes we can have this enumeration of our calls to the Redis API done separately from this PR.
@oshadmi Let's put this in our backlog, WDYT? I think it shouldn't be that hard, and is a good stat to add to our `FT.INFO` response. It's useful for both the users and us.
Regarding the type consistency - GREAT. Then we can merge this as is in that case.

**Review by raz-mon — APPROVED** [review]

**Review by GuyAv46 — APPROVED** [review]

**Review by oshadmi — APPROVED** [review]

**Pull request merged**